hibernate-core/src/main/java/org/hibernate/graph/internal/AbstractGraph.java

```diff
@@ -124,7 +124,12 @@ public AttributeNodeImplementor<?> addAttributeNode(AttributeNodeImplementor<?>
 	@Override
 	@SuppressWarnings("unchecked")
 	public <AJ> AttributeNodeImplementor<AJ> findAttributeNode(String attributeName) {
-		PersistentAttribute<? super J, ?> attribute = managedType.findAttributeInSuperTypes( attributeName );
+		PersistentAttribute<?, ?> attribute = managedType.findAttributeInSuperTypes( attributeName );
+
+		if ( attribute == null ) {
+			attribute = managedType.findSubTypesAttribute( attributeName );
```

Comment:
Please revert this. Implicitly looking into subtypes is not an option. Users will have to explicitly cast the graph to "treat" it and can then access subtype attributes.

Reply:
Hello, 
Sorry for the delay. I do understand. I have the impression that the difficulty here is that if I remove this as well as the [graph merge below](https://github.com/hibernate/hibernate-orm/pull/9088#discussion_r1808912130), the subgraphs on which a type is specified will be ignored. 
This is because when fields are traversed in StandardEntityGraphTraversalStateImpl, all subtypes attributes are represented as if they belonged to the parent class. 
Today on [this line](https://github.com/hibernate/hibernate-orm/blob/6584eb3d53c32ed041d40a17a9b1ed169e2ee347/hibernate-core/src/main/java/org/hibernate/sql/results/internal/StandardEntityGraphTraversalStateImpl.java#L87), when we have sub-graphs associated with sub-types, we ignore them. 



In the case below, for example, the ‘**course_sub_graph** sub-graph will be ignored:
```

	@NamedEntityGraph(
			name = "exemple",
			subgraphs = {
					@NamedSubgraph(
							name = "course_sub_graph", 
							type = PayingCourse.class,
							attributeNodes = { @NamedAttributeNode("moneyReceiver") }
					)
			}
	)
	@DiscriminatorValue("student")
	public static class Student  {
		@Id
		Long id;
		@OneToMany(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
		List<Course> courses;
	}

	@Entity(name = "Course")
	@DiscriminatorColumn(name = "course_type", discriminatorType = DiscriminatorType.STRING)
	public static abstract class Course {
		@Id
		Long id;
		String name;
	}

	@Entity(name = "FreeCourse")
	public static class FreeCourse extends Course {
	}

	@Entity(name = "PayingCourse")
	@DiscriminatorValue("paying")
	public static class PayingCourse extends Course {

		@ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
		Teacher moneyReceiver;
	}


```

Reply:
@Asanio06 so what you're saying is that the explicit `class = PayingCourse.class` _is _ doing a "treat".

The spec is for this stuff is _extremely_ hard to understand, and I wasn't around when it was added, so I don't know for sure what it's supposed to mean, however, section 10.3.3 and the following Javadoc would appear to support your interpretation:

```java
public @interface NamedSubgraph {

    /**
     * (Required) The name of the subgraph as referenced from a
     * NamedAttributeNode element.
     */
    String name();

    /**
     * (Optional) The type represented by this subgraph. The element
     * must be specified when this subgraph is extending a definition
     * on behalf of a subclass.
     */
    Class<?> type() default void.class;

    /** 
     * (Required) The list of the attributes of the class that must
     * be included. If the named subgraph corresponds to a subclass
     * of the class referenced by the corresponding attribute node,
     * then only subclass-specific attributes are listed.
     */
    NamedAttributeNode[] attributeNodes();
}
```

Reply:
@gavinking  
Yes, that's exactly it, it's not very well documented. 
I have the impression that to support this inheritance functionality on subgraphs, it is necessary to go to the attributes of child classes on these function. 
If not, [currentGraphContext.findAttributeNode](https://github.com/hibernate/hibernate-orm/blob/6584eb3d53c32ed041d40a17a9b1ed169e2ee347/hibernate-core/src/main/java/org/hibernate/sql/results/internal/StandardEntityGraphTraversalStateImpl.java#L61) will always return null. And if we don't merge the different sub-graphs, **currentGraphContext** will correspond to only 1 of the sub-graphs and the others will be ignored.

I have the impression that the other way requires changing the way you browse an entity's attributes to find out if they are to be fetched.

Reply:
Well, also read section 10.3.3 of the spec, which tends to imply that you are *only* allowed to specify directly-declared members of `PayingCourse.class` in that `@NamedSubgraph`.

So perhaps that rule helps avoid searching through hierarchies? 

<blockquote>

The `NamedSubgraph` annotation is used to
further define an attribute node. It is referenced by its name from the
`subgraph` or `keySubgraph` element of a `NamedAttributeNode` element.

The `name` element is the name used to
reference the subgraph from a `NamedAttributeNode` definition. In the
case of entity inheritance, multiple subgraph elements have the same
name.

The `type` element must be specified when the
subgraph corresponds to a subclass of the entity type corresponding to
the referencing attribute node.

The `attributeNodes` element lists attributes
of the class that must be included. If the subgraph corresponds to a
subclass of the class referenced by the corresponding attribute node,
only subclass-specific attributes are listed.
</blockquote>

Also:

<blockquote>
The `subgraph` element is used to refer to a
`NamedSubgraph` specification that further characterizes an attribute
node corresponding to a managed type (entity or embeddable). The value
of the `subgraph` element must correspond to the `name` used for the
subgraph in the `NamedSubgraph` element. If the referenced attribute is
an entity which has entity subclasses, there may be more than one
`NamedSubgraph` element with this name, and the `subgraph` element is
considered to refer to all of these.
</blockquote>

Thanks for looking into this stuff, but the way, @Asanio06, I know it's a hole we have.

Reply:
@gavinking  You're right, these details really help me in my reasoning.
Does this mean that to solve our problem we could potentially make SubGraphImpl contain a ‘subclassSubgraphs’ variable like RootGraphImpl?
So if I create 3 subgraphs of the same name but different types on the same AttributeNode, I'll have 1 subgraph that has a link to the other subgraphs. 

Is this correct?

Reply:
I *think* so?

Reply:
@gavinking @beikov Thanks for all your help
I'm going to work on the changes and I'll get back to you as soon as it's done.

Reply:
Awesome, thanks again man!

Reply:
Hello @gavinking @beikov ,
I removed that.
Here's a new version. 
This time if I consider that each attribute node has only 1 subgraph which itself can have subclassSubgraphs.
This change generates numerous impacts that I've tried to manage as well as possible.

In the StandardEntityGraphTraversalStateImpl, I now use the declaringType of the fetchable to determine whether I should use the graph directly or whether I should use a subclassSubgraph. This is the way I've found to avoid unnecessary attribute node traversals.